src/internal_modules/roc_core/mov_histogram.h

```diff
@@ -0,0 +1,123 @@
+/*
+ * Copyright (c) 2023 Roc Streaming authors
+ *
+ * This Source Code Form is subject to the terms of the Mozilla Public
+ * License, v. 2.0. If a copy of the MPL was not distributed with this
+ * file, You can obtain one at http://mozilla.org/MPL/2.0/.
+ */
+
+//! @file roc_core/mov_histogram.h
+//! @brief Rolling window moving histogram.
+
+#ifndef ROC_CORE_MOV_HISTOGRAM_H_
+#define ROC_CORE_MOV_HISTOGRAM_H_
+
+#include "roc_core/array.h"
+#include "roc_core/iarena.h"
+#include "roc_core/panic.h"
+#include "roc_core/ring_queue.h"
+
+namespace roc {
+namespace core {
+
+//! @brief A class that implements a rolling window moving histogram.
+//! The MovHistogram class maintains a histogram of values within a specified window
+//! length. It divides the range of values into a specified number of bins and updates the
+//! histogram as new values are added and old values are removed from the window.
+//! @tparam T The type of values to be histogrammed.
+
+template <typename T> class MovHistogram {
+public:
+    //! @brief Constructs a moving histogram.
+    //! @param arena Memory arena for dynamic allocations.
+    //! @param value_range_min The minimum value of the range to be histogrammed.
+    //! @param value_range_max The maximum value of the range to be histogrammed.
+    //! @param num_bins The number of bins in the histogram. Each bin represents a
+    //! subrange of the value range.
+    //! @param window_length The length of the moving window. Only values within this
+    //! window are considered in the histogram.
+
+    MovHistogram(IArena& arena,
+                 T value_range_min,
+                 T value_range_max,
+                 size_t num_bins,
+                 size_t window_length)
+        : value_range_min_(value_range_min)
+        , value_range_max_(value_range_max)
+        , num_bins_(num_bins)
+        , window_length_(window_length)
+        , ring_buffer_(arena, window_length)
+        , bins_(arena)
+        , valid_(false) {
+        if (num_bins == 0 || window_length == 0 || value_range_min >= value_range_max) {
+            roc_panic("mov histogram: number of bins and window length must be greater "
+                      "than 0 and value_range_min must be less than value_range_max");
+        }
+
+        bin_width_ = (value_range_max - value_range_min) / static_cast<T>(num_bins);
```

Comment:
Hey @adrianopol,

Since `num_bins` is initialized as `size_t`, it needs to be cast to type `T`. Please let me know if this is incorrect.

Reply:
This is correct, I just mean that C/C++ already do this implicit promotion (see https://en.cppreference.com/w/cpp/language/implicit_conversion): if at least on of operands in an arithmetic expression have type float (or double) then the whole expression will have type float (or double).

If T = float, then without casting we have `(float - float) / size_t` == `float / size_t` == `float`.

If T = `short`, we have `(short - short) / size_t` == `short / size_t` == `size_t`. Value (max - min) is always positive, so we have no problem with loosing sign here.

You may leave your variant as is, it is correct too.

Reply:
If T is `short`, then `(value_range_max - value_range_min) / num_bins` is size_t, and assigning size_t to bin_width_ would trigger a warning.

If T is `ssize_t` (signed version of size_t), then `(value_range_max - value_range_min) / num_bins` would trigger a warning.

So I think we need to keep the cast here.

BTW we usually use c-style casts or constructor-style casts instead of static_cast, but that's not a big deal.